debug handler maintain through decomposition

Gasoonjia opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141612

Add checks in the ao numberic debugger to guard the debug handle consistency between aten op decomposition

Differential Revision: [D66517480](https://our.internmc.facebook.com/intern/diff/D66517480/)

## BC-breaking note

### `torch.ao.quantization.pt2e.graph_utils.get_control_flow_submodules` is No Longer a Public API

We aim to make all functions under `torch.ao.quantization.pt2e.graph_utils` private. This update marks `get_control_flow_submodules` as a private API.

Currently, only `bfs_trace_with_node_process` remains public, but it will be marked private in the next release.

## BC-breaking note

### `torch.ao.quantization.pt2e.graph_utils.get_control_flow_submodules` is No Longer a Public API

We aim to make all functions under `torch.ao.quantization.pt2e.graph_utils` private. This update marks `get_control_flow_submodules` as a private API. If you have to or want to continue using `get_control_flow_submodules`, please make a private call by using `_get_control_flow_submodules`.

Currently, only `bfs_trace_with_node_process` remains public, but it will be marked private in the next release.

**Example:**
Version 2.6:
```python
>>> from torch.ao.quantization.pt2e.graph_utils import get_control_flow_submodules
  ```

Version 2.7:
```python
>>> from torch.ao.quantization.pt2e.graph_utils import get_control_flow_submodules
ImportError: cannot import name 'get_control_flow_submodules' from 'torch.ao.quantization.pt2e.graph_utils'
>>> from torch.ao.quantization.pt2e.graph_utils import _get_control_flow_submodules # Note: Use _get_control_flow_submodules for private access
```

Gasoonjia committed:
debug handler maintain through decomposition

Add checks in the ao numberic debugger to guard the debug handle consistency between aten op decomposition

Differential Revision: [D66517480](https://our.internmc.facebook.com/intern/diff/D66517480/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "debug handler maintain through decomposition"


Add checks in the ao numberic debugger to guard the debug handle consistency between aten op decomposition

Differential Revision: [D66517480](https://our.internmc.facebook.com/intern/diff/D66517480/)

[ghstack-poisoned]

Gasoonjia commented:
#suppress-bc-linter

Gasoonjia committed:
Update on "debug handler maintain through decomposition"


Add checks in the ao numberic debugger to guard the debug handle consistency between aten op decomposition

Differential Revision: [D66517480](https://our.internmc.facebook.com/intern/diff/D66517480/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "debug handler maintain through decomposition"


Add checks in the ao numberic debugger to guard the debug handle consistency between aten op decomposition

Differential Revision: [D66517480](https://our.internmc.facebook.com/intern/diff/D66517480/)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/pt2e/test_numeric_debugger.py:
```diff
@@ -128,29 +160,67 @@ def test_re_export_preserve_handle(self):
+            ep_copy = copy.copy(ep)
+            ep_copy = ep_copy.run_decompositions()
+            self._assert_each_node_has_debug_handle(ep_copy)
+            pre_decomp_to_debug_handle_map = (
```
you mean `post`?

Gasoonjia replied:
It is `pre`. The map key is the pre-decompose operator name, and value is the debug handle in the post debug handle. Maybe I need to have a better name haha

jerryzh168 replied:
oh I see, makes sense

Gasoonjia committed:
Update on "debug handler maintain through decomposition"


Add checks in the ao numberic debugger to guard the debug handle consistency between aten op decomposition

Differential Revision: [D66517480](https://our.internmc.facebook.com/intern/diff/D66517480/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "debug handler maintain through decomposition"


Add checks in the ao numberic debugger to guard the debug handle consistency between aten op decomposition

Differential Revision: [D66517480](https://our.internmc.facebook.com/intern/diff/D66517480/)

[ghstack-poisoned]

albanD commented on docs/source/conf.py:
```diff
@@ -305,6 +305,7 @@
     "node_arg_is_weight",
     "return_arg_list",
     # torch.ao.quantization.pt2e.graph_utils
+    "bfs_trace_with_node_process",
```
Adding new entries in this file is not ok.
Please make sure that all the new public APIs that you add are properly documented and show up on the website.

Gasoonjia replied:
Thanks for the feedback. Any example I can use for documentation? I noticed that all public APIs under `torch.ao.quantization.pt2e.graph_utils` are in this file and (looks like) there's not any doc for `graph_utils`.

Gasoonjia replied:
If allowed, we can first add `bfs_trace_with_node_process` here, and have another PR for all doc stuff under `graph_utils`.

jerryzh168 replied:
graph_utils shouldn't be public API, I'd recommend to make this private (add `_`)

Gasoonjia committed:
Update on "debug handler maintain through decomposition"


Add checks in the ao numberic debugger to guard the debug handle consistency between aten op decomposition

Differential Revision: [D66517480](https://our.internmc.facebook.com/intern/diff/D66517480/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "debug handler maintain through decomposition"


Add checks in the ao numberic debugger to guard the debug handle consistency between aten op decomposition

Differential Revision: [D66517480](https://our.internmc.facebook.com/intern/diff/D66517480/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "debug handler maintain through decomposition"


Add checks in the ao numberic debugger to guard the debug handle consistency between aten op decomposition

Differential Revision: [D66517480](https://our.internmc.facebook.com/intern/diff/D66517480/)

[ghstack-poisoned]
